include/zephyr/zbus/zbus.h

```diff
@@ -0,0 +1,555 @@
+	/** Dynamic channel observer list. Represents the channel's observers list, it can be empty
+	 * or have listeners and subscribers mixed in any sequence. It can be changed in runtime.
+	 */
+	sys_slist_t runtime_observers;
+#endif /* CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE  */
+};
+
+/**
+ * @brief Type used to represent an observer.
+ *
+ * Every observer has an representation structure containing the relevant information.
+ * An observer is a code portion interested in some channel. The observer can be notified
+ * synchronously or asynchronously and it is called listener and subscriber respectively.
+ * The observer can be enabled or disabled during runtime by change the enabled boolean
+ * field of the structure. The listeners have a callback function that is executed by the
+ * bus with the index of the changed channel as argument when the notification is sent.
+ * The subscribers have a message queue where the bus enqueues the index of the changed
+ * channel when a notification is sent.
+ *
+ * @see zbus_obs_set_enable function to properly change the observer's enabled field.
+ *
+ */
+struct zbus_observer {
+#if defined(CONFIG_ZBUS_OBSERVER_NAME) || defined(__DOXYGEN__)
+	/** Observer name. */
+	const char *name;
+#endif
+	/** Enabled flag. Indicates if observer is receiving notification. */
+	bool enabled;
+	/** Observer message queue. It turns the observer into a subscriber. */
+	struct k_msgq *queue;
+
+	/** Observer callback function. It turns the observer into a listener. */
+	void (*callback)(const struct zbus_channel *chan);
+};
+
+/** @cond INTERNAL_HIDDEN */
+
+#if defined(CONFIG_ZBUS_ASSERT_MOCK)
+#define _ZBUS_ASSERT(_cond, _fmt, ...)                                                             \
+	({                                                                                         \
+		if (!(_cond)) {                                                                    \
+			LOG_ERR(_fmt, ##__VA_ARGS__);                                              \
+			return -EFAULT;                                                            \
+		}                                                                                  \
+	})
+#else
+#define _ZBUS_ASSERT(_cond, _fmt, ...) __ASSERT(_cond, _fmt, ##__VA_ARGS__)
+#endif
+
+#if defined(CONFIG_ZBUS_CHANNEL_NAME)
+#define ZBUS_CHANNEL_NAME_INIT(_name) .name = #_name,
+#else
+#define ZBUS_CHANNEL_NAME_INIT(_name)
+#endif
+
+#if defined(CONFIG_ZBUS_OBSERVER_NAME)
+#define ZBUS_OBSERVER_NAME_INIT(_name) .name = #_name,
+#define _ZBUS_OBS_NAME(_obs)	       (_obs)->name
+#else
+#define ZBUS_OBSERVER_NAME_INIT(_name)
+#define _ZBUS_OBS_NAME(_obs) ""
+#endif
+
+#if CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE > 0
+#define ZBUS_RUNTIME_OBSERVERS_LIST_INIT                                                           \
+	.runtime_observers = SYS_SLIST_STATIC_INIT(&_name.runtime_observers),
+#else
+#define ZBUS_RUNTIME_OBSERVERS_LIST_INIT /* No user data */
```

Comment:
Not anymore. It would be better to leave the developers to allocate the `user data`. The `user data` field only points to it during the channel's definition macro. Doing this allows each channel to have independent `user data`, which was impossible before. The user data sample illustrates how to do that.

Reply:
I meant the comment. It does not seem to be related to the `ZBUS_RUNTIME_OBSERVERS_LIST_INIT` macro. It looks like some copy-paste remnant.

Reply:
Indeed. I will remove that.